lib/awful/tag.lua

```diff
@@ -74,6 +75,66 @@ local function raw_tags(scr)
     return tmp_tags
 end
 
+local function custom_layouts(self)
+    local cls = tag.getproperty(self, "custom_layouts")
+
+    -- Allow custom layouts to be listed until GC-ed.
+    if not cls then
+        cls = {}
+        tag.setproperty(self, "custom_layouts", cls)
+    end
+
+    return cls
+end
+
+-- Create a custom list if none is set or update the one to replace `from` by
+-- `to`. This is necessary for layouts defined as functions or dynamic layouts.
+local function update_layouts(self, from, to)
+    if not to then return end
+
+    alayout = alayout or require("awful.layout")
+    local override = tag.getproperty(self, "layouts")
+
+    local pos = from and gtable.hasitem(override or {}, from) or nil
+
+    -- There is an override and the layout template is part of it, replace by
+    -- the instance.
+    if override and pos and from ~= to then
+        assert(type(pos) == 'number')
+        override[pos] = to
+        self:emit_signal("property::layouts")
+        return
+    end
+
+    -- Only add to the custom_layouts and preserve the ability to globally
+    -- set the layouts.
+    if override and not pos then
+        table.insert(override, to)
+        self:emit_signal("property::layouts")
+        return
+    end
+
+    pos = from and gtable.hasitem(alayout.layouts, from) or nil
+
+    -- The new layout is part of the global layouts. Fork the list.
+    if pos and from ~= to then
+        local cloned = gtable.clone(alayout.layouts, false)
+        table.insert(cloned, to)
+        self.layouts = cloned
+        return
+    end
+
+    if pos then return end
```

Comment:
This is only reached if `from == to`, right? If so, this can be removed if the early return is added at the beginning of the function.